### ISSUE TITLE
[ASRangeController] Some nodes may be orphaned after insertions and deletions

### ISSUE THREAD

**Issue by ERussel** [facebook/AsyncDisplayKit#791]
I can confirm that the issue is still present. For example, when i call reloadData for collectionView, invisible nodes would not be deallocated. As i understand from ASRangeHandlerRender code, invisible nodes are added to global working window to activate offscreen rendering, but i don't see any workaround to remove them from superview in DELETE_NODES call.

**Comment by appleguy** [comment]
@ERussel Thanks for reporting this!  Just to confirm, you're not using ASNetworkImageNode or ASBasicImageDownloader, correct?

This should be an easy fix if the cause you highlighted is correct.  @levi, @nguyenhuy could this have been impacted by recent refactoring?  I would be pretty surprised if no one else had reported this if it is a longstanding issue, but it's certainly possible.

**Comment by levi** [comment]
@appleguy indeed could be a regression from the recent changes. I don't have much time this week to take a look, Sunday at the earliest. @nguyenhuy if you're able to check this out earlier that would be awesome!

**Comment by nguyenhuy** [comment]
I can take a look tomorrow :)

**Comment by ERussel** [comment]
Thank you for attention to this issue.

In fact, it is very old. I can confirm it on 1.1.1 version. I made temporarily fix for my project

- (void)clearInvisibleNodes{
    UIWindow *window = [[ASRangeHandlerRender class] performSelector:NSSelectorFromString(@"workingWindow")];
    for (UIView *view in [window subviews]) {
        if ([view isKindOfClass:[_ASDisplayView class]]) {
            ASDisplayNode *node = [view performSelector:@selector(asyncdisplaykit_node)];
            if ([node isKindOfClass:[GiphyCollectionViewNode class]]) {
                if (node.isNodeLoaded) {
                    if (node.layerBacked) {
                        [node.layer removeFromSuperlayer];
                    } else {
                        [node.view removeFromSuperview];
                    }
                }
            }
        }
    }
}

I combine call to this method with reloadData. I checked with Profile Leak Instrument and it seems that invisible node release properly now.

P.S. I know about leak related to ASNetworkImage when used with ASBasicImageDownloader and i am confident that bugs are different.

**Comment by appleguy** [comment]
Great clarifying information, much appreciate it.On Oct 29, 2015, at 10:36 AM, Ruslan <notifications@github.com> wrote:Thank you for attention to this issue.In fact, it is very old. I can confirm it on 1.1.1 version. I made temporarily fix for my project


(void)clearInvisibleNodes{
UIWindow *window = [[ASRangeHandlerRender class] performSelector:NSSelectorFromString(@"workingWindow")];
for (UIView *view in [window subviews]) {
    if ([view isKindOfClass:[_ASDisplayView class]]) {
        ASDisplayNode *node = [view performSelector:@selector(asyncdisplaykit_node)];
        if ([node isKindOfClass:[GiphyCollectionViewNode class]]) {
            if (node.isNodeLoaded) {
                if (node.layerBacked) {
                    [node.layer removeFromSuperlayer];
                } else {
                    [node.view removeFromSuperview];
                }
            }
        }
    }
}
}
I combine call to this method with reloadData. I checked with Profile Leak Instrument and it seems that invisible node release properly now.P.S. I know about leak related to ASNetworkImage when used with ASBasicImageDownloader and i am confident that bugs are different.—Reply to this email directly or view it on GitHub.

**Comment by nguyenhuy** [comment]
> In fact, it is very old. I can confirm it on 1.1.1 version. I made temporarily fix for my project
> I combine call to this method with reloadData. I checked with Profile Leak Instrument and it seems that invisible node release properly now.

This is probably a silly question, but this means we don't need to investigate this issue anymore, right?

**Comment by ERussel** [comment]
I mean the issue exists at least from 1.1.1 version. I made a hack fix example for 2 reason: ensure that the bug is related to invisible nodes and to avoid leaks in my project's current version. But i think the issue needs to be fixed in a better way in the AsyncDisplayKit library.

**Comment by nguyenhuy** [comment]
Ok, thanks for the explanation. I will spend some time on this issue soon.

**Comment by ERussel** [comment]
Ok, thank you!

**Comment by nguyenhuy** [comment]
@ERussel I should have asked you these questions earlier: what version of ASDK are you using? Can you reproduce this issue on one of our examples?

I've been trying to identify the leaks by periodically reloading the collection view in our ASCollectionView example, but Leaks Instrument has shown no problems so far, on both `master` and `v1.2.3`.

**Comment by ERussel** [comment]
Thank you for response.

I will try to reproduce the leak on ASCollection view example. But do you have any answer to question: "When invisible nodes gets removed from shared window?"

**Comment by eanagel** [comment]
I can confirm we are seeing this in our application (on 1.2.3) with ASTableView as well.

**Comment by eanagel** [comment]
@nguyenhuy I'm currently looking at this issue for my project and I think I could have something PR-ready later today. I want to make sure we aren't duplicating effort. Are you in the middle of this one already?

**Comment by nguyenhuy** [comment]
@eanagel: My incompetent skills can't get me any further. I was about to ask for your help. Please go a head.

**Comment by nguyenhuy** [comment]
@ERussel

> "When invisible nodes gets removed from shared window?"

The backing store (view or layer, but most likely view in this case) of invisible nodes are removed from the shared window by `ASRangeHandlerRender`, but the nodes themselves are still retained by `ASDataController`.

**Comment by ERussel** [comment]
@nguyenhuy 

> The backing store (view or layer, but most likely view in this case) of invisible nodes are removed from the shared window by ASRangeHandlerRender, but the nodes themselves are still retained by ASDataController. They should be deallocated when the data controller deletes them, or when it reloads data.

Yes, but as i see from code, corresponding method of ASRangeHandlerRender is called only when visible range changes due to collection view/table view scroll action.

**Comment by nguyenhuy** [comment]
I'm sorry for my unclear explanation in the previous comment. I have just updated it to be more precise. 

Yes, it is true that ASRangeHandlerRender is called (by ASRangeController) when visible range changes, because that's when some nodes fell out of the working range and their backing store should be removed from the working window.

**Comment by appleguy** [comment]
@eanagel @nguyenhuy @ERussel does anyone have a test case for this?  I am ready to debug now (I have another meeting at 3 though).  I know the range controller code and can probably fix it, but need a reproduction case.

**Comment by appleguy** [comment]
I just hacked the kittens app to call -reloadData when the button in the navigation bar is tapped.  I also removed the imageNode because of the known retain cycle cause by the ASBasicImageDownloader.  It is not showing any leaked memory.

**Comment by nguyenhuy** [comment]
@eanagel said he can reproduce this in his app and is already working on a diff.

**Comment by appleguy** [comment]
Here's the trace, with the dip caused at the point I hit reload.  The memory ramp up before then was from when I started scrolling down and more things were being added to the working range.  So far this is indicating no reproduction of the bug described.  Are the cases you guys are seeing also without any possibility of retain cycles?  Have you checked with the Instruments cycle detector?

<img width="655" alt="screen shot 2015-10-30 at 2 03 14 pm" src="https://cloud.githubusercontent.com/assets/565251/10858010/02b6126a-7f0f-11e5-92e9-beeb25b12d45.png">

**Comment by appleguy** [comment]
yeah, @eanagel pinged me and suggested I take a look.  I think he's AFK right now.

**Comment by nguyenhuy** [comment]
Same thing for me. I tried to reload Kitten and ASCollectionView examples and saw memory went up then back down just like the image above.

**Comment by eanagel** [comment]
Hey guys. I've started a new issue this is a little more specific to the problem I think I'm seeing #797. Issues arise when you either tear down a ASTable/CollectionView or when you are doing a lot of inserts/deletes. I'm doing lots of all the above which might be why this is a more serious issue in my case.

**Comment by appleguy** [comment]
@ERussel - Hey man, could you help me reproduce this?  I had plenty of time to fix it this weekend, but had to stop looking at it because everything I tried couldn't make it happen.  I totally believe that it is possible, and have ideas for fixes, but can't test whether they work and so can't really make progress.

If you could post a diff / .zip file / PR / or even just written instructions on how to modify any of the examples/ projects to reproduce, that would be fantastic!

I'm going to land my PR which makes some changes that hopefully allow us to reproduce with only a few additional tweaks:  https://github.com/facebook/AsyncDisplayKit/pull/801

cc @eanagel - I think it is fine to wait until after your super urgent deadline coming up this week.  Even after that, I am more than comfortable taking over the fix for this issue.  If you do have time to help, the only thing I need is reproduction steps.

**Comment by eanagel** [comment]
@appleguy Let me see if I repro with my virtualization example project...

**Comment by ERussel** [comment]
@appleguy @eanagel @nguyenhuy 

Sorry for delay. I made a simple project to reproduce the issue. You can find it here https://github.com/ERussel/TestColoredCollectionView

To reproduce the issues make the following steps:

1. Profile with Allocations Instrument.
2. Scroll 3-4 rows.
3. Switch to green.
4. Scroll 3-4 row.
5. Switch to blue.
6. Sort allocations in Profile window.

Normally, you should see only allocated BlueCell objects, but due to bug you will see also allocated RedCell and GreenCell objects. 

I hope, i could explain! Thank you for time spending to this issue.

**Comment by appleguy** [comment]
Thanks so much!  This should be enough.  I have a full meeting day at work tomorrow, but will definitely start debugging on Wednesday.

**Comment by eanagel** [comment]
Ok, I have one for you too! Check out my sample here: https://github.com/eanagel/ASDKVirtualizationDemo

1. Profile with allocations instrument
2. Click the "Virtualization Demo" button.
3. Scroll for a little while through the date list
4. Back back out to the root view controller.
5. Take a look in allocations, searching for objects "AS*". You should see several instances of ASTextCellNode & friends there.

Of course the ASTableView has been torn down and you shouldn't see any AS* objects laying around at all. If you dig a bit you'll find them hanging around in the workingWindow (Which is why they don't show up in leaks - they have a valid root.)

**Comment by appleguy** [comment]
There is a side discussion with several interesting comments here: https://github.com/facebook/AsyncDisplayKit/issues/797 - merging it into this one.

**Comment by appleguy** [comment]
@ERussel this should now be fixed after #830 - please reopen if you still see a relevant number of leaks.

Looking at @eanagel's test project, there is actually one instance of ASTextNode (but zero instances of ASCellNode, etc) that is orphaned per intense usage session of the view.  The amount of memory is incredibly tiny.  However, this is still not acceptable, so feel free to follow #828 for the complete 100% fix of that (I have confirmed that even in the diff's rough state, it does eliminate /all/ orphaned / leaked elements in every test case I have).

If you have a test case that is orphaning more than just one object per creation / destruction cycle of the collection or table, please let me know.  Still though, most likely the best fix will be #828.
